### PR TITLE
content: Add embedded async Rust article

### PR DESCRIPTION
**Pull request by bahildebrand**

### CODE REVIEW

File: _drafts/embedded-async-rust.md [hunk 1]
```diff
@@ -0,0 +1,681 @@
...
+
+```rust
+if LED_BLINK_ACTIVE.load(Ordering::SeqCst) {
+    led.toggle();
+}
+```
+
+Next, we have a small change in our `blinky` task. We've added a check to see if
+the LED should be blinking. If it should, we toggle the LED. If not, we do
+nothing. This is a simple way to pause the blinking of the LED when the button
```
Comment (ejohnso49) [inline _drafts/embedded-async-rust.md:406]:
```suggestion
nothing. This is a simple way to pause the blinking of the LED when we press the button.
```

File: _drafts/embedded-async-rust.md [hunk 2]
```diff
@@ -0,0 +1,681 @@
...
+We now understand how Rust futures work under the hood, and the role an executor
+plays in managing and driving futures to completion. The question remains,
+why would we go through all this trouble? What benefit do we gain over a
+traditional preemptive RTOS or a bare-metal super loop? The answer is I/O!
+
+One of the biggest benefits of async Rust is smaller code size. A traditional
+RTOS will require a stack for each task. Rust futures only need to keep track
+of variables that are used across await points. This means that the memory
+needed for an async task can be much smaller than a traditional RTOS task, as
+small 10s of bytes for less complex tasks! This means we can spin up more tasks
```
Comment (sarfata) [inline _drafts/embedded-async-rust.md:613]:
```suggestion
small as 10s of bytes for less complex tasks! This means we can spin up more tasks
```

File: _drafts/embedded-async-rust.md [hunk 3]
```diff
@@ -0,0 +1,681 @@
...
+polled when it gets its next turn.
+
+## Benefits of Async Rust on Embedded Systems
+
+We now understand how Rust futures work under the hood, and the role an executor
+plays in managing and driving futures to completion. The question remains,
+why would we go through all this trouble? What benefit do we gain over a
+traditional preemptive RTOS or a bare-metal super loop? The answer is I/O!
+
+One of the biggest benefits of async Rust is smaller code size. A traditional
```
Comment (sarfata) [inline _drafts/embedded-async-rust.md:609]: code size? seems like you mean RAM requirements here? Maybe both are true ;)

File: img/author/blake.jpg
Comment (sarfata) [inline img/author/blake.jpg:1]: 🤯

**Review by sarfata — APPROVED** [review]
Super awesome stuff. Great article!

File: _drafts/embedded-async-rust.md [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+hood. This can make it difficult to debug issues when they arise. An overloaded
+task that is starving others in the runtime can be difficult to diagnose. This
+isn't helped by the fact that debug symbols can get a little weird when using
+async Rust. This is because stack unwinding isn't quite what you'd
+expect when looking because of the way futures are polled and scheduled.
+
+Another downside is the lack of priorities introduced by the cooperative
+scheduler. To its credit, Embassy has a pretty cool solution to this problem.
+Using the priorities assigned to interrupts, multiple executors can be run in
+said interrupts[^6]. This allows for a form of priority-based scheduling, but
```
Comment (ejohnso49) [inline _drafts/embedded-async-rust.md:637]:
```suggestion
Using the priorities assigned to interrupts, an executor is run in each of these interrupts[^6]. This allows for a form of priority-based scheduling, but
```

**Review by ejohnso49 — APPROVED** [review]
Looks good!

File: img/embedded-rust-async/cooperative.png
Comment (noahp) [inline img/embedded-rust-async/cooperative.png:1]: If you use the excalidraw vscode plugin, you can edit in vscode and save them as png/svg with the source embedded, so it's extremely painless to tweak them:

https://marketplace.visualstudio.com/items?itemName=pomdtr.excalidraw-editor

Example:
https://github.com/memfault/interrupt/blob/master/img/stacks/stack-growth-double-link.excalidraw.svg

File: _drafts/embedded-async-rust.md (hunk 3 above)
Comment (bahildebrand) [reply _drafts/embedded-async-rust.md:609]: I sure did, oof

File: img/embedded-rust-async/cooperative.png
Comment (bahildebrand) [reply img/embedded-rust-async/cooperative.png:1]: Great tip thanks!

**Pull request merged**